docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+process templates.*
+
+* Parameter definitions include metadata which describes the purpose of the parameter.  Since parameter definitions are part of the template,
+  there is no need to indicate which template they correspond to.
+
+*It should be possible to validate templates and template parameters, both values and the schema.*
+
+* Template objects are subject to standard api validation.
+
+*It should also be possible to validate and view the output of the substitution process.*
+
+* The `/processedTemplates` api returns the result of the substitution process, which is itself a Template object that can be validated.
+
+*It should be possible to generate forms for parameterized templates, as discussed in #4210 and #6487.*
+
+* Parameter definitions provide metadata that allows for the construction of form-based UIs to gather parameter values from users.
+
+*It shouldn't be inordinately difficult to evolve templates. Thus, strategies such as versioning and encapsulation should be 
+encouraged, at least by convention.*
+
+* Templates can be versioned via annotations on the template object.
+
+## Alternatives
+
+### Mustache other generic templating language
+
+[Mustache](https://mustache.github.io/) offers a semi-standard generic way to do templating with substitution.  The primary 
+limitation is that the syntax is not valid json (unless all parameters are strings and delimited by quotes), and parameter 
+values must be provided separately.  This means Mustache templates could not be stored and referenced as first class objects.
+
+Other templating systems surveyed had similar limitations:  separated parameters from the template, did not provide generator 
+capabilities, and/or did not use a syntax which could easily be rendered as a Kubernetes API object.
+
+#### Stream transformers
+As defined [here](https://github.com/kubernetes/kubernetes/pull/14993) and [here](https://github.com/kubernetes/kubernetes/pull/14918).
+
+The primary limitation here seems to be that it is a client-side proposal, generators would have to be provided in the client, meaning third party clients/web consoles/etc would have a hard time taking advantage of this mechanism.  It also doesn’t provide a way to create and store new generic templates, you must have a generator defined and then provide appropriate parameters for that generator, so it doesn’t seem like it fills the same space as the templates in this proposal.
+
+Lastly it doesn’t provide a way to list/describe parameters consumed by the template, or provide default values in the template itself, making it hard to build a UI experience around.
+
+#### Deployment Manager
+
+As described [here](https://github.com/kubernetes/deployment-manager)
+
+* Utilizes multiple files with different syntax to define an application topology.
+* Does not define a first class API object.
+* Does not appear to natively offer generation logic for producing parameter values.
+
```

Comment:
> Utilizes multiple files with different syntax to define an application topology.

this example uses 4 files including both yaml and jinja formats: https://github.com/kubernetes/deployment-manager/tree/master/examples/wordpress

the docs also indicate templates can be written in python.  none of the examples/formats match k8s api objects/syntax.

> Does not define a first class API object.

see above, what DM defines/consumes is not a first class k8s api object.  Perhaps there's some confusion in what i meant by api object, I did not mean that DM doesn't define an api for itself, I meant it does not define api objects that are consumed by k8s.  I can clarify this statement.

> Does not appear to natively offer generation logic for producing parameter values.

I understand DM has parameters, i don't see native logic for defining how a parameter value should be generated (eg generate a random password).  Presumably it's possible to make that happen via jinja or python, but I wouldn't consider that native support within DM, hence my statement.  Again, I can clarify this to indicate value generation is deferred to jinja/python.

If I am still not understanding the design of DM, I'm afraid I'm going to need you to be more explicit about what is inaccurate/what I have misunderstood.